### PR TITLE
Check user access to service on sign in

### PR DESCRIPTION
**Pull request by steventux**
## Context

We would like to check user roles as an authorisation step on the `check-records` part of the service.
The DfE Signin API [exposes an endpoint](https://github.com/DFE-Digital/login.dfe.public-api#get-user-access-to-service) which replies with user roles for a given user, service and organisation.

## Changes

- Call the DfE Sign in API _get user access to service_ endpoint as part of the DSI login journey.
- Authorise user based on presence of valid role for service, valid role codes are stored in `ENV["DFE_SIGN_IN_API_ROLE_CODES"]`
- Create a `DsiUserSession` record on each successful login (except when bypass is active). This record contains the role and org info the user authenticated with.

### Guidance to review

Local overrides are necessary to make this PR work. Specifically the env vars
```
DFE_SIGN_IN_API_BASE_URL
DFE_SIGN_IN_API_SECRET
DFE_SIGN_IN_API_AUDIENCE
DFE_SIGN_IN_API_ROLE_CODES
```
I have edited the review app keyvault with working values.

### Link to Trello card

https://trello.com/c/mTFXWXo7/166-dfe-signin-role-spike
https://trello.com/c/7vJeIbez/185-store-more-things-about-the-user

### Checklist

- [x] Attach to Trello card
- [x] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/lib/check_records/dfe_sign_in.rb [hunk 1]
```diff
@@ -13,5 +13,53 @@ def self.bypass_active?
...
+
+      def client
+        @client ||=
+          Faraday.new(
+            url: ENV.fetch("DFE_SIGN_IN_API_BASE_URL"),
+            request: {
+              timeout: TIMEOUT_IN_SECONDS
+            }
+          ) do |faraday|
+            faraday.request :authorization, "Bearer", token
```
Comment (steventux) [inline app/lib/check_records/dfe_sign_in.rb]: This isn't right,  we need JWT Auth here https://github.com/DFE-Digital/login.dfe.public-api#get-user-access-to-service

**Comment by steventux** [comment]
[DfE Signin 'role' spike](https://trello.com/c/mTFXWXo7/166-dfe-signin-role-spike)

**Comment by steventux** [comment]
[Store more things about the user](https://trello.com/c/7vJeIbez/185-store-more-things-about-the-user)

**Review by gpeng — APPROVED** [review]

**Review by malcolmbaig — APPROVED** [review]
LGTM

**Pull request merged**